File: FrontEnd/lib/domain/usecases/plants/get_user_plants_use_case.dart
/// @file get_user_plants_use_case.dart
/// @description Implementación del caso de uso para obtener las plantas del usuario.
/// @module Plants
/// @layer Domain
library;

import '../../entities/plant.dart';
import '../../interfaces/usecases/plants/i_get_user_plants_use_case.dart';
import '../../repositories/i_plant_repository.dart';

/// [implements] IGetUserPlantsUseCase
/// [dependencies] IPlantRepository
class GetUserPlantsUseCase implements IGetUserPlantsUseCase {
  final IPlantRepository _repository;
  const GetUserPlantsUseCase({required IPlantRepository repository})
      : _repository = repository;

  @override
  Future<List<Plant>> execute() => _repository.getUserPlants();
}

File: BackEnd/src/domain/usecases/plants/GetUserPlantsUseCase.ts
/**
 * @file GetUserPlantsUseCase.ts
 * @description Caso de uso para obtener la lista de plantas del usuario autenticado.
 * Calcula nextWatering actualizado antes de mapear al DTO de respuesta.
 * @module Plants
 * @layer Domain
 *
 * @implements {IGetUserPlantsUseCase}
 * @injectable
 * @dependencies IPlantRepository, IPlantMapper
 */

import { injectable, inject } from 'inversify';
import { TYPES } from '../../../core/types.js';
import type { IGetUserPlantsUseCase } from '../../interfaces/usecases/plants/IGetUserPlantsUseCase.js';
import type { IPlantRepository } from '../../repositories/IPlantRepository.js';
import type { IPlantMapper } from '../../../data/IMappers/IPlantMapper.js';
import type { PlantResponseDTO } from '../../dtos/plants/plant-response.dto.js';

/**
 * Obtiene todas las plantas activas del usuario.
 *
 * @implements {IGetUserPlantsUseCase}
 * @injectable
 * @dependencies IPlantRepository, IPlantMapper
 */
@injectable()
export class GetUserPlantsUseCase implements IGetUserPlantsUseCase {
  constructor(
    @inject(TYPES.IPlantRepository) private readonly plantRepo: IPlantRepository,
    @inject(TYPES.IPlantMapper)     private readonly mapper: IPlantMapper,
  ) {}

  /**
   * Retorna las plantas del usuario, con nextWatering recalculado si es necesario.
   *
   * @param userId — Id del usuario autenticado.
   * @returns Lista de PlantResponseDTO.
   */
  async execute(userId: string): Promise<PlantResponseDTO[]> {
    const plants = await this.plantRepo.findByUserId(userId);
    return plants.map((plant) => this.mapper.toResponseDTO(plant));
  }
}
